---
title: "BeePollen_IRIA_Project"
author: "Minuka Hewapathirana"
date: "08/11/2023"
output: html_document
fontsize: 11pt
geometry: margin=1in
urlcolor: blue
header-includes: \usepackage{xcolor}
---

```{r, global_options, eval=TRUE, echo=FALSE, cache=TRUE}
library(knitr)
opts_chunk$set(tidy.opts=list(width.cutoff=80), tidy = TRUE, fig.align='center',
               cache=FALSE, collapse=TRUE, echo=FALSE, eval=FALSE, include=FALSE,
               message=FALSE, quietly=TRUE, results='hide', warn.conflicts=FALSE, 
               warning=FALSE)
```
Project Name Setting:
```{r}
analysis <- "BeePollen"
```

Source our custom R scripts:    
For this we will use the rprojroot package to set the directory structures. This will help us when finding our files to source functions. We specify ours is an RStudio project. The root object contains a function that will help us locate our package R files regardless of our current working directory.
```{r sourcing_my_functions, echo=FALSE, eval=TRUE, include=FALSE, cache=TRUE}
library("rprojroot")
root <- rprojroot::is_rstudio_project

scriptsPath <- root$make_fix_file(".")("R")
scripts  <- dir(root$find_file("R", path = root$find_file()))
scriptsList <- paste(scriptsPath, scripts, sep = "//")
lapply(scriptsList, source)

scriptsPyPath <- root$make_fix_file(".")("pythonScripts")
dir.create(scriptsPyPath, showWarnings = TRUE, recursive = TRUE)
scriptsPy <- dir(root$find_file(scriptsPyPath, path = root$find_file()))
scriptsPyList <- paste(scriptsPyPath, scriptsPy, sep = "//")

imagePath <- root$make_fix_file(".")(paste("r_environments", analysis, sep = "/"))
dir.create(imagePath, showWarnings = TRUE, recursive = TRUE)
images <- dir(root$find_file(imagePath, path = root$find_file())) # if there is already a data file in this directory if not running this for the first time
imagesList <- paste(imagePath, images, sep = "/")
imagesList
load(imagesList) # you need to be specific if you have more than one image saved in this directory
```

Record the path to the environment images directory:
```{r importChptImage, echo=TRUE, eval=TRUE, include=FALSE, cache=TRUE}
library("rprojroot")
# sharedPath <- root$make_fix_file(".")("PIRL_working_directory")
# dir.create(sharedPath, showWarnings = TRUE, recursive = TRUE)
# sharedPathDirs <- dir(root$find_file(sharedPath, path = root$find_file()))
# sharedPathAn <- paste(sharedPath, analysis, sep = "/")
# dir.create(sharedPathAn, showWarnings = TRUE, recursive = TRUE)

sharedPathAn <- "/isilon/CFIA-ACIA_projects/girouxeml/hewapathiranam/PIRL_working_directory/BeePollenAnalysis"
```

```{r}
databasePath <- root$make_fix_file(".")(paste("Databases", analysis, sep = "/"))
dir.create(databasePath, showWarnings = TRUE, recursive = TRUE)
databasesPathDir <- dir(root$find_file(databasePath, path = root$find_file()))

refSeqDB <- paste(databasePath, "refSeq", sep = "/")
gbDB <- paste(databasePath, "genbank", sep = "/")
ensemblDB <- paste(databasePath, "ensembl", sep = "/")
viruSiteDB <- paste(databasePath, "viruSite", sep = "/")
subfolder_names <- c("/refSeq", "/genbank", "/ensembl", "/viruSite")
for (j in 1:length(subfolder_names)){
  folder <- dir.create(paste0(databasePath, subfolder_names[j]), showWarnings = TRUE, recursive = FALSE)
}

# species within genus genomes 
generaGenomes <- paste(databasePath, "all_genomes_under_target_genera", sep = "/")
refSeqDBGenus <- paste(generaGenomes, "refSeq", sep = "/")
gbDBGenus <- paste(generaGenomes, "genbank", sep = "/")
ensemblDBGenus <- paste(generaGenomes, "ensembl", sep = "/")
viruSiteDBGenus <- paste(generaGenomes, "viruSite", sep = "/")
for (j in 1:length(subfolder_names)){
  folder <- dir.create(paste0(generaGenomes, subfolder_names[j]), showWarnings = TRUE, recursive = TRUE)
}
```

```{r}
imageA <- "aps_test.RData" # should provide this variable in initial chunk where image directory is specified - but concern is that there may be the need to divide and create different files within the same anaylsis....
save.image(paste(imagePath, imageA, sep = "/")) # you need to bget more specific if you have more than one image saved in this directory
```

Reference genomes to obtain for target species, read csv:
```{r}
library("rprojroot")
library("data.table")
targetCsv <- "targetPathogens.csv"
targetCsvPath <- root$find_file(targetCsv)
targetMeta <- data.table::fread(targetCsvPath, sep = "auto", header = TRUE)
```

Need to obtain reference genomes for the target pathogens in the target Meta table:
```{r}
library("biomartr")
# set default timeout setting from 60s to at least 3000000 s before running any retrieval functions:
options(timeout = 30000000)
# List available database:
biomartr::listNCBIDatabases(db = "all")
# also try BarcodingR: https://besjournals.onlinelibrary.wiley.com/doi/full/10.1111/2041-210X.12682
```

This is for interactive runs of the chunk - further down is the qsub.    
To-do: specify that this is "interactive" in chunk options, set up chunk with setting to specify if running interactive script or qsub mode, then have the chunk run depending on the setting.    
    
RefSeq Genome Check:
```{r}
library("biomartr")
# set default timeout setting from 60s to at least 3000000 s before running any retrieval functions:
options(timeout = 30000000)
for(i in 1:nrow(targetMeta)){
  res <- biomartr::is.genome.available(db = "refseq", organism = targetMeta$SpeciesName[i])
  targetMeta$genRefSeq[i] <- res
  }
refSeqGens <- targetMeta[targetMeta$genRefSeq == "TRUE",]
for(i in 1:nrow(refSeqGens)){
  res <- biomartr::is.genome.available(db = "refseq", organism = refSeqGens$SpeciesName[i], details = TRUE)
  refSeqGens$assmblyAccn[i] <- res
}
save.image(paste(imagePath, imageA, sep = "/"))
```

This is for interactive runs of the chunk - further down is the qsub.    
To-do: specify that this is "interactive" in chunk options, set up chunk with setting to specify if running interactive script or qsub mode, then have the chunk run depending on the setting.    

RefSeq Genome Download of most recent version. Set up as a background job (Don't copy results - just double-check download folder has all genomes downloaded):
```{r}
options(timeout = 300000000)
library(biomartr)
for(i in 1:nrow(refSeqGens)){
  biomartr::getGenomeSet(db = "refseq", refSeqGens$SpeciesName[i], reference = FALSE, release = NULL, 
                         clean_retrieval = TRUE, gunzip = FALSE, update = FALSE, path = refSeqDB, 
                         assembly_type = "toplevel")
  biomartr::getAssemblyStats(db = "refseq", refSeqGens$SpeciesName[i], reference = FALSE, 
                             type = "download", 
                             #path = paste(refSeqDB, "/genomeassembly_stats", sep = ""))
                             path = refSeqDB)
  warnings()
}
```

This is for interactive runs of the chunk - further down is the qsub.    
To-do: specify that this is "interactive" in chunk options, set up chunk with setting to specify if running interactive script or qsub mode, then have the chunk run depending on the setting.    

Save env image:
```{r}
save.image(paste(imagePath, imageA, sep = "/"))
```

This is for interactive runs of the chunk - further down is the qsub.    
To-do: specify that this is "interactive" in chunk options, set up chunk with setting to specify if running interactive script or qsub mode, then have the chunk run depending on the setting.    

Ensembl Genome Check:    
    
** Note - if there is an error returned:    
Error in if (!is.taxid(organism)) { :     
  missing value where TRUE/FALSE needed    
      
This error not a script error, this is because there are no ENSEMBL geneomes available - if you look at the table , "targetMeta" you'll see that all the species have "FALSE" for the ensembl column. This error will occur in the next chunk as well.
```{r}
for(i in 1:nrow(targetMeta)){
  res <- biomartr::is.genome.available(db = "ensembl", organism = targetMeta$SpeciesName[i])
  targetMeta$genEMBL[i] <- res
}
ensemblGens <- targetMeta[targetMeta$genEMBL == "TRUE",]
for(i in 1:nrow(ensemblGens)){
  res <- biomartr::is.genome.available(db = "ensembl", organism = ensemblGens$SpeciesName[i], details = TRUE)
  ensemblGens$assmblyAccn[i] <- res
}
```

This is for interactive runs of the chunk - further down is the qsub.    
To-do: specify that this is "interactive" in chunk options, set up chunk with setting to specify if running interactive script or qsub mode, then have the chunk run depending on the setting.    

Ensembl Genome Retrieval: (Run as background job)
```{r}
library(biomartr)
for(i in 1:nrow(ensemblGens)){
  biomartr::getGenomeSet(db = "ensembl", ensemblGens$SpeciesName[i], reference = FALSE, release = NULL,
                         clean_retrieval = TRUE, gunzip = FALSE, update = FALSE, path = ensemblDB,
                         assembly_type = "toplevel")
    biomartr::getAssemblyStats(db = "ensembl", ensemblGens$SpeciesName[i], reference = FALSE, 
                             type = "download", 
                             path = ensemblDB)
  warnings()
}
```

This is for interactive runs of the chunk - further down is the qsub.    
To-do: specify that this is "interactive" in chunk options, set up chunk with setting to specify if running interactive script or qsub mode, then have the chunk run depending on the setting.    

Save env image:
```{r}
save.image(paste(imagePath, imageA, sep = "/"))
```

This is for interactive runs of the chunk - further down is the qsub.    
To-do: specify that this is "interactive" in chunk options, set up chunk with setting to specify if running interactive script or qsub mode, then have the chunk run depending on the setting.    
    
GenBank Genome Check:
```{r}
for(i in 1:nrow(targetMeta)){
  res <- biomartr::is.genome.available(db = "genbank", organism = targetMeta$SpeciesName[i])
  targetMeta$genDB[i] <- res
}

gbGens <- targetMeta[targetMeta$genDB == "TRUE",]
for(i in 1:nrow(gbGens)){
  res <- biomartr::is.genome.available(db = "genbank", organism = gbGens$SpeciesName[i], details = TRUE)
  gbGens$assmblyAccn[i] <- res
}
```

This is for interactive runs of the chunk - further down is the qsub.    
To-do: specify that this is "interactive" in chunk options, set up chunk with setting to specify if running interactive script or qsub mode, then have the chunk run depending on the setting.    

GenBank Genome Retrieval: (Run as background job)
```{r}
library(biomartr)
for(i in 1:nrow(gbGens)){
  biomartr::getGenomeSet(db = "genbank", gbGens$SpeciesName[i], reference = FALSE, release = NULL,
                         clean_retrieval = TRUE, gunzip = FALSE, update = FALSE, path = gbDB,
                         assembly_type = "toplevel")
    biomartr::getAssemblyStats(db = "genbank", gbGens$SpeciesName[i], reference = FALSE, 
                             type = "download", 
                             path = gbDB)
  warnings()
}
```

Genome Check QSub - GenBank, RefSeq, Ensembl    
    
Note: to run qsub with more than 1 node, instead of running the bash.sh script in the prefix directory, run the command script directly with qsub specifications, i.e.,:    
$ qsub -pe smp 8 -cwd A_Target_Genome_Check1.sub    
Note that we don't include the "-S /bin/bash " because we're running an rscript, not a bash script  for this qsub.
```{r}
prefix <- "A_Target_Genome_Check"
cmd <- paste('cmdA <- "source /isilon/cfia-ottawa-fallowfield/users/girouxeml/prog/anaconda3/bin/activate && conda activate /isilon/cfia-ottawa-fallowfield/users/girouxeml/prog/anaconda3/envs/rstudio" \n',
             'system(cmdA) \n',
             "load('", paste(imagePath, "/", imageA, sep = ""), "') \n",
             'library("biomartr") \n',
             'options(timeout = 30000000) \n',
             'for(i in 1:nrow(targetMeta)){ \n',
                  'res <- biomartr::is.genome.available(db = "genbank", organism = targetMeta$SpeciesName[i]) \n', 
                  'targetMeta$genDB[i] <- res \n',      
                  '}\n',
             '\n',
             'gbGens <- targetMeta[targetMeta$genDB == "TRUE",] \n',
             'for(i in 1:nrow(gbGens)){\n',
                  'res <- biomartr::is.genome.available(db = "genbank", organism = gbGens$SpeciesName[i], details = TRUE) \n',
                  'gbGens$assmblyAccn[i] <- res \n',
                  '} \n',
             '\n',
             "save.image('", paste(imagePath, "/", imageA, sep = ""), "') \n",
             
             'for(i in 1:nrow(targetMeta)){ \n',
                  'res <- biomartr::is.genome.available(db = "ensembl", organism = targetMeta$SpeciesName[i]) \n',
                  'targetMeta$genEMBL[i] <- res \n',
                  '}\n',
             'ensemblGens <- targetMeta[targetMeta$genEMBL == "TRUE",] \n',
             '#for(i in 1:nrow(ensemblGens)){ \n',
                  '#res <- biomartr::is.genome.available(db = "ensembl", organism = ensemblGens$SpeciesName[i], details = TRUE) \n',
                  '#ensemblGens$assmblyAccn[i] <- res # Must comment out this line or script will fail because embl likely will not have genomes and leave this here so that I do not foget why\n',
                  '#} \n',
             '\n',
             'for(i in 1:nrow(targetMeta)){ \n',
                  'res <- biomartr::is.genome.available(db = "refseq", organism = targetMeta$SpeciesName[i]) \n',
                  'targetMeta$genRefSeq[i] <- res \n',
                  '} \n',
             'refSeqGens <- targetMeta[targetMeta$genRefSeq == "TRUE",] \n',
             'for(i in 1:nrow(refSeqGens)){ \n',
                  'res <- biomartr::is.genome.available(db = "refseq", organism = refSeqGens$SpeciesName[i], details = TRUE) \n',
                  'refSeqGens$assmblyAccn[i] <- res \n',
                  '}\n',
             '\n',
             "save.image('", paste(imagePath, "/", imageA, sep = ""), "') \n",
             '\n',
             'cmdB <- "source /isilon/cfia-ottawa-fallowfield/users/girouxeml/prog/anaconda3/bin/deactivate && conda deactivate" \n',
             'system(cmdB) \n',
             sep = "")

MakeRQsubs(cmd, prefix)
```




Genome Retrieval QSub - Genbank and RefSeq (EMBL ommitted since no genomes avaialbe and including it would break script)    
Use this chunk to run the above 2 chunks as qsubs on the biocluster instead of interactively.   
    
Note: to run qsub with more than 1 node, instead of running the bash.sh script in the prefix directory, run the command script directly with qsub specifications, i.e.,:    
$ qsub -pe smp 8 -cwd B_Retrieve_Target_Genomes1.sub    
Note that we don't include the "-S /bin/bash " because we're running an rscript, not a bash script  for this qsub.
```{r, echo=TRUE, eval=FALSE, include=TRUE, tidy=TRUE}
prefix <- "B_Retrieve_Target_Genomes"
cmd <- paste('cmdA <- "source /isilon/cfia-ottawa-fallowfield/users/girouxeml/prog/anaconda3/bin/activate && conda activate /isilon/cfia-ottawa-fallowfield/users/girouxeml/prog/anaconda3/envs/rstudio" \n',
             'system(cmdA) \n',
             "load('", paste(imagePath, "/", imageA, sep = ""), "') \n",
             'library("biomartr") \n',
             'options(timeout = 30000000) \n',
             'for(i in 1:nrow(gbGens)){ \n',
                  'biomartr::getGenomeSet(db = "genbank", gbGens$SpeciesName[i], 
                        reference = FALSE, release = NULL, clean_retrieval = TRUE, 
                        gunzip = FALSE, update = FALSE, path = gbDB, 
                        assembly_type = "toplevel") \n',
                  '\n',
                  'biomartr::getAssemblyStats(db = "genbank", gbGens$SpeciesName[i], 
                        reference = FALSE, type = "download", path = gbDB) \n',
                  'warnings() \n',
                  '} \n',
             '\n',
             "save.image('", paste(imagePath, "/", "fetchedGens_", imageA, sep = ""), "') \n",
             '\n',
             'for(i in 1:nrow(refSeqGens)){ \n',
                  'biomartr::getGenomeSet(db = "refseq", refSeqGens$SpeciesName[i], 
                        reference = FALSE, release = NULL, clean_retrieval = TRUE, 
                        gunzip = FALSE, update = FALSE, path = refSeqDB, 
                        assembly_type = "toplevel") \n',
                  '\n',
                  'biomartr::getAssemblyStats(db = "refseq", refSeqGens$SpeciesName[i], 
                        reference = FALSE, type = "download", path = refSeqDB) \n',
                  'warnings() \n',
                  '} \n',
             '\n',
             "save.image('", paste(imagePath, "/fetchedGens_", imageA, sep = ""), "') \n",
             '\n',
             'cmdB <- "source /isilon/cfia-ottawa-fallowfield/users/girouxeml/prog/anaconda3/bin/deactivate && conda deactivate" \n',
             'system(cmdB) \n',
             sep = "")


MakeRQsubs(cmd, prefix)
```

```{r}
load(paste(imagePath, "/", "fetchedGens_", imageA, sep = ""))
```
Save env image:
```{r}
save.image(paste(imagePath, imageA, sep = "/"))
```
** Over here!!! 28Mar2023 To-do: need to repeat genome fetches but for genus level - may include target species if there is a genome for it available, but many won't - will still allow us to see close relatives and later expand to in-silico primer design, etc. 

Species with no Genomes In GenBank and RefSeq:
```{r}
library("data.table")
otherDBs <- targetMeta[targetMeta$genDB == "FALSE" & targetMeta$genRefSeq == "FALSE" & targetMeta$genEMBL == "FALSE",]
```
```{r}
otherDBs$SpeciesName
colnames(otherDBs)
otherDBs[Kingdom %in% c("Bacteria", "Fungi"),]
otherDBs[Kingdom %like% "Bacteria",]
otherDBs[Kingdom != "Virus"]
```


Viral Genomes not found using any databases compatible with biomartr:
```{r}
virusGens <- targetMeta[targetMeta$Kingdom == "Virus"]
virusGens$SpeciesName
```
Prepare for python function "nearest_genomes.py"
```{r}
library("data.table")
generaList <- unique(targetMeta$Genus)
nearestGensTbl <- as.data.table(generaList)
```

*** Over here!!!! 29Mar2023 Go back to the python nearest genomes script - need to get the accessions available for each genus:
```{r}

```

```{r}
library("reticulate")
nearestGenPyPath <- paste(scriptsPyPath, "nearest_genomes.py", sep = "/")
# use_python(pythonPath)
source_python(nearestGenPyPath)

for(i in 1:nrow(nearestGensTbl)){
  accns_r <- nearest_genomes(nearestGensTbl$generaList[i])
  nearestGensTbl$accns[i] <- paste(accns_r, collapse = ", ")
}
```



### Finding genomes for the other targets:    
Xanthomonas translucens pv. Undulosa    2021    NCBI Assembly ASM1730177v1    GenBank   GCA_017301775.1    RefSeq   GCF_017301775.1    Modifier   XtLr8 (strain)   URL:    ncbi.nlm.nih.gov/data-hub/genome/?taxon=487909    
    
Pseudomonas syringae pv. striafaciens   NCBI Assembly ASM369965v1   GenBank   GCA_003699655.1   RefSeq    GCF_003699655.1   Modifier  ICMP 4418 (strain)    URL:      ncbi.nlm.nih.gov/data-hub/genome/?taxon=235276    
    
Pseudomonas syringae pv. coronafaciens   NCBI Assembly ASM369995v1   GenBank   GCA_003699955.1   RefSeq    GCF_003699695.1   Modifier  ICMP 10052 (strain)    URL:      ncbi.nlm.nih.gov/data-hub/genome/?taxon=235275    
    
Urocystis agropyri NCBI TaxID 265355 - no complete genome assembly available - need to look more into possible resources for this species   
    
Tilletia tritici   NCBI Assembly Tca517_3.0   GenBank   GCA_004334575.1   RefSeq    N/A   Modifier  517 (strain)    URL:      ncbi.nlm.nih.gov/data-hub/genome/?taxon=13290    
            
Cephalosporium gramineum   NCBI TaxID 259995 - no complete genome assembly available - need to look more into possible resources for this species. Current scientific name [Hymenula] cerealis    
    
Diaporthe phaseolorum var. sojae   NCBI TaxID 165439 - no complete genome assembly available - need to look more into possible resources for this species.    
        
Peronospora manshurica   NCBI TaxID 86334 - no complete genome assembly available - need to look more into possible resources for this species.    



### Ideas:    

Looking at viruSite, genomes appear to be gathered from other NCBI databases - consider if biomartr script can access these?    
biomartr::listNCBIDatabases(db = "all"):    

*** Note - all the viral genomes we are missing in targetMeta are available in the virusite.org database - consider programmatic way to retrieve them    
To do:    
- write chunk to query the viruSite website to download genomes - in the mean time, download manually     
- in mean time - need to add info to meta that viusite genomes are included 
Add virus genomes to metadata, and their paths - as these were manually downloaded from ViruSite.org:
```{r}
# list files in path

```

             





*** Over here!!!! Setting right python path loaded to continue with the script. Not necessary?
```{r}
# find environment of the conda path for this script - this one is for using the rstudio env:
envPath <- root$make_fix_file(".")("../../prog/anaconda3/envs/rstudio")
dir.create(envPath, showWarnings = TRUE, recursive = TRUE)
envDirs <- dir(root$find_file(envPath, path = root$find_file()))
envBinPath <- root$make_fix_file(".")(paste(envPath, "bin", sep = "/"))
envBinList <- dir(root$find_file(envBinPath, path = root$find_file()))
pythonPath <- paste(envBinPath, "python", sep = "/")
```


```{r}
library("reticulate")
nearestGenPyPath <- paste(scriptsPyPath, "nearest_genomes.py", sep = "/")
# use_python(pythonPath)
source_python(nearestGenPyPath)

for(i in 1:nrow(nearestGensTbl)){
  accns_r <- nearest_genomes(nearestGensTbl$generaList[i])
  nearestGensTbl$accns[i] <- paste(accns_r, collapse = ", ")
}
```


Ensembl Genome Retrieval: (Run as background job)
```{r}
library(biomartr)
for(i in 1:nrow(nearestGensTbl)){
  biomartr::getGenomeSet(db = "ensembl", nearestGensTbl$accns[i], reference = FALSE, release = NULL,
                         clean_retrieval = TRUE, gunzip = FALSE, update = FALSE, path = ensemblDBGenus,
                         assembly_type = "toplevel")
    biomartr::getAssemblyStats(db = "ensembl", nearestGensTbl$accns[i], reference = FALSE, 
                             type = "download", 
                             path = ensemblDBGenus)
  warnings()
}

for(i in 1:nrow(targetMeta)){
  res <- biomartr::is.genome.available(db = "genbank", organism = targetMeta$SpeciesName[i])
  targetMeta$genDB[i] <- res
}

gbGens <- targetMeta[targetMeta$genDB == "TRUE",]
for(i in 1:nrow(gbGens)){
  res <- biomartr::is.genome.available(db = "genbank", organism = gbGens$SpeciesName[i], details = TRUE)
  gbGens$assmblyAccn[i] <- res
}

library(biomartr)
for(i in 1:nrow(gbGens)){
  biomartr::getGenomeSet(db = "genbank", gbGens$SpeciesName[i], reference = FALSE, release = NULL,
                         clean_retrieval = TRUE, gunzip = FALSE, update = FALSE, path = gbDB,
                         assembly_type = "toplevel")
    biomartr::getAssemblyStats(db = "genbank", gbGens$SpeciesName[i], reference = FALSE, 
                             type = "download", 
                             path = gbDB)
  warnings()
}
```


Obtain genomes from all species within a genus available from database:
```{r}
getKingdoms(db = "refseq")
getGroups(db = "refseq", kingdom = "fungi")
```



Using the new NCBI's datasets API to get the nearest available reference genomes for a given taxon or organism name.
Input: TAX_ID or Name
Output:
    1- The nearest organism with available reference genomes
    2- Accessions of the reference genomes
Requirements:
pip install --upgrade ncbi-datasets-pylib

*** over here!!!
Okay - so now need to figure out how to run and collect the output for this python code for each row in column of metadata... 22nov2022:
```{r}
organism_tax_or_name = targetMeta$Genus
```




Trying to find a way to get the genomes of all species withing a genus. Some ideas:     
- taxize github.com/ropensci/taxize, https:://reslp.github.io/blog/Download-Taxinfo-with-R/
- taxonkit - need to install in conda env, not an R package
- BarcodingR package - but need to see how useful this would be for us




```{r}
BiocManager::install("edgeR")
install.packages(c("taxize", "myTAI", "plyr", "usethis"))
library("taxize")
library("myTAI")
library("plyr")
taxize::use_entrez()
usethis::edit_r_environ()
ENTREZ_KEY='7f4be04d1855def2fe2aa44c9e2b94bc9a09'
myTAI::taxonomy(organism = targetMeta$Genus[1],
                db = "ncbi",
                output = "classification")




```


```{r}
rawPath <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/PIRL_working_directory/advancedPreScreening/data/raw/17Dec2021_HortonAllLongAmpRun2_cat.fastq"
rawSummary <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/PIRL_working_directory/advancedPreScreening/data/raw/sequencing_summary_17Dec2021_HortonAllLongAmpRun2_cat.txt"
```
 
### Basecalling    
    
### Error correction    
Vaser et al, 2017, "Fast and accurate de novo genome assembly from long uncorrected reads" 
- used RACON to skip error-correction and get consensus seqs with a SIMD-accelerated, partial-order alignment-based, stand-alone consensus module https://genome.cshlp.org/content/27/5/737
    
### consensus-generation   

"A workflow for accurate metabarcoding using nanopore MinION sequencing"
https://besjournals.onlinelibrary.wiley.com/doi/full/10.1111/2041-210X.13561

"build error-corrected consensus sequences"
Python pipeline "ASHURE" and "OPTICS density-based clustering"for:
- data processing
consensus building
clustering - use reference genomes, primer indices marking each sample, or spatially-related sequence information
taxonomic assignment

"ASHURE" compared to C3POa workflow for consensus error correction of nanopore reads

see what our median accuracies are
see how many successfully identified species can be retrieved from the mock sample


https://www.sciencedirect.com/science/article/pii/S1872497321000326
clustering method "NGSpeciesID" tp generate accurate consensus sequence for species identification
- consensus seqs constructed from a multiple seq alignment software:
- seq reads are clustered based on seq homology with the aim of generating a single consensus seq for every source of input DNA
- for species id, this means that one consensus seq should be generated per distinct taxon present within a sample
- then sompared against a reference database to identify the species




really cool method to increase accuracy is RCA - rolling circle amplification so that replicated sequneces can be used to build concensus sequences with an accuracy of up to 99.5%

metabarcoding and nanopore sequencing could allow researchers wto generate barcode sequence data for community samples in the field without the need to transport or ship samples to a lab.

gff <- read.table(file = gffPathPr, 
                  sep = "\t", header = F, quote = "", comment.char="#", fill = T)
                  
metadata <-  read.table(paste(sharedPathAn, metadataFileAlternate, sep = ""),
                        sep = ",", header = TRUE, comment.char = "", quote = "", as.is = TRUE)
                        
                        
for(i in 1:nrow(metadataAssembly)){
    kmerdatTmp <- fread(metadataAssembly$kmerGenieDatPath[i], sep = "auto", header = TRUE)
    setkey(kmerdatTmp, genomic.kmers)
    key(kmerdatTmp)
    maxGenomicKmers <- max(kmerdatTmp$genomic.kmers)
    metadataAssembly$BestKmerGenie[i] <- kmerdatTmp[.(maxGenomicKmers), .(k)]
}



i <- 1
for(i in 1:length(listABySS_fac_res)){
    test[[i]] <- fread(listABySS_fac_res[i], sep = "auto", header = TRUE)
    test[[i]]$ABySSfastq <- basename(test[[i]]$name)
    print(nrow(test[[i]]))
}

blastnTmp <- fread(metadataAssemblySubTemp$LawiiMitoBlastPath[i], sep = "auto", header = FALSE)

cazyLawiAll <- fread(metadataAssembly$domtbloutPath[7], sep = "auto", header = TRUE, fill=TRUE)

cazy analysis RstudioScript-June2017_LachnellulaSpp_assemblies.Rmd line 2055


library(reshape2)
library(data.table)
dbCANList <- metadataAssembly$ParsedmtblPath
i <- 1
test <- lapply(dbCANList, fread, sep = "auto", header = FALSE)
namesCols <- c("family.hmm", "hmm.length", "query.id", "query.length", "e.val",
               "hmm.start", "hmm.end", "query.start", "query.end", "coverage")
for(i in 1:length(test)){
  setnames(test[[i]], 1:10, namesCols)
}

dbCAN_tempFile <- fread(paste(sharedPathAn, "dbCAN_CAZySubset_AllRefs.csv", sep = ""),
                        sep="auto", header = TRUE)
                        
library(data.table)
metadataAssemblies <- fread(paste(sharedPathAn, "Lachnellula_genomes_MetadataAssembly.csv", sep = ""),
                            sep = "auto", header = TRUE)
metadataAssemblies[, V1:=NULL]




